File: opencv_example/scripts/hough_transform_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from rcl_interfaces.msg import SetParametersResult
import cv2
import cv_bridge
import numpy as np


class HoughTransformNode(Node):
    def __init__(self):
        super().__init__('hough_transform_node')

        self.blue_thres_ = self.declare_parameter('blue_thres', 130).value
        self.erode_size_ = self.declare_parameter('erode_size', 1).value
        self.dilate_size_ = self.declare_parameter('dilate_size', 1).value
        self.hough_rho_res_ = self.declare_parameter('hough_rho_res', 10).value
        self.hough_theta_res_ = self.declare_parameter('hough_theta_res', 0.05).value
        self.hough_threshold_ = self.declare_parameter('hough_threshold', 10).value
        self.hough_min_length_ = self.declare_parameter('hough_min_length', 20).value
        self.hough_max_gap_ = self.declare_parameter('hough_max_gap', 50).value
        self.add_on_set_parameters_callback(self.param_update)

        self.bridge_ = cv_bridge.CvBridge()
        self.sub_image_ = self.create_subscription(Image, 'image', self.recv_image, 1)

        cv2.namedWindow('Raw Image', cv2.WINDOW_AUTOSIZE)
        cv2.namedWindow('Blue Image', cv2.WINDOW_AUTOSIZE)
        cv2.namedWindow('Thres Image', cv2.WINDOW_AUTOSIZE)
        cv2.namedWindow('Erode Image', cv2.WINDOW_AUTOSIZE)
        cv2.namedWindow('Dilate Image', cv2.WINDOW_AUTOSIZE)
        cv2.namedWindow('Canny Image', cv2.WINDOW_AUTOSIZE)
        cv2.namedWindow('Lines Image', cv2.WINDOW_AUTOSIZE)

    def recv_image(self, msg):
        # Convert raw image from ROS image message into a numpy array
        raw_img = self.bridge_.imgmsg_to_cv2(msg, msg.encoding)

        cv2.imshow('Raw Image', raw_img)
        cv2.waitKey(1)

        # Split RGB image into its three separate channels
        split_images = cv2.split(raw_img)

        # Extract the blue channel into its own grayscale image
        blue_image = split_images[0]

        cv2.imshow('Blue Image', blue_image)
        cv2.waitKey(1)

        # Apply binary threshold to create a binary image where white pixels correspond to high blue values
        _, thres_img = cv2.threshold(blue_image, self.blue_thres_, 255, cv2.THRESH_BINARY)

        cv2.imshow('Thres Image', thres_img)
        cv2.waitKey(1)

        # Apply erosion to clean up noise
        erode_kernel = np.ones(shape=(self.erode_size_, self.erode_size_), dtype=np.uint8)
        erode_img = cv2.erode(thres_img, erode_kernel)

        cv2.imshow('Erode Image', erode_img)
        cv2.waitKey(1)

        # Apply dilation to expand regions that passed the erosion filter
        dilate_kernel = np.ones(shape=(self.dilate_size_, self.dilate_size_), dtype=np.uint8)
        dilate_img = cv2.dilate(erode_img, dilate_kernel)

        cv2.imshow('Dilate Image', dilate_img)
        cv2.waitKey(1)

        # Apply Canny edge detection to reduce the number of points passed to Hough Transform
        canny_img = cv2.Canny(dilate_img, 1, 2)
        cv2.imshow('Canny Image', canny_img)

        # Run Probabilistic Hough Transform algorithm to detect line segments
        line_segments = cv2.HoughLinesP(image=canny_img,
                                        rho=self.hough_rho_res_,
                                        theta=self.hough_theta_res_,
                                        threshold=self.hough_threshold_,
                                        minLineLength=self.hough_min_length_,
                                        maxLineGap=self.hough_max_gap_)

        # Draw detected Hough lines onto the raw image for visualization
        if line_segments is not None:
            for seg in line_segments:
                x1, y1, x2, y2 = seg[0]
                cv2.line(raw_img, (x1, y1), (x2, y2), (0, 0, 255))

        cv2.imshow('Lines Image', raw_img)
        cv2.waitKey(1)

    def param_update(self, parameters):
        for p in parameters:
            if p.name == 'blue_thres':
                self.blue_thres_ = p.value
            elif p.name == 'erode_size':
                self.erode_size_ = p.value
            elif p.name == 'dilate_size':
                self.dilate_size_ = p.value
            elif p.name == 'hough_rho_res':
                self.hough_rho_res_ = p.value
            elif p.name == 'hough_theta_res':
                self.hough_theta_res_ = p.value
            elif p.name == 'hough_threshold':
                self.hough_threshold_ = p.value
            elif p.name == 'hough_min_length':
                self.hough_min_length_ = p.value
            elif p.name == 'hough_max_gap':
                self.hough_max_gap_ = p.value

        result = SetParametersResult()
        result.successful = True
        return result


if __name__ == '__main__':
    rclpy.init()
    node_instance = HoughTransformNode()
    rclpy.spin(node_instance)

File: opencv_example/src/hough_transform_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <opencv2/opencv.hpp>
#include <cv_bridge/cv_bridge.hpp>

namespace opencv_example {

  class HoughTransformNode : public rclcpp::Node {
    public:
      HoughTransformNode()
      : rclcpp::Node("hough_transform_node")
      {
        blue_thres_ = declare_parameter<int>("blue_thres", 130);
        erode_size_ = declare_parameter<int>("erode_size", 1);
        dilate_size_ = declare_parameter<int>("dilate_size", 1);
        hough_rho_res_ = declare_parameter<int>("hough_rho_res", 10);
        hough_theta_res_ = declare_parameter<double>("hough_theta_res", 0.05);
        hough_threshold_ = declare_parameter<int>("hough_threshold", 10);
        hough_min_length_ = declare_parameter<int>("hough_min_length", 20);
        hough_max_gap_ = declare_parameter<int>("hough_max_gap", 50);
        param_cb_ = add_on_set_parameters_callback(std::bind(&HoughTransformNode::param_update, this, std::placeholders::_1));

        sub_image_ = create_subscription<sensor_msgs::msg::Image>("image", 1, std::bind(&HoughTransformNode::recv_image, this, std::placeholders::_1));

        cv::namedWindow("Raw Image", cv::WINDOW_AUTOSIZE);
        cv::namedWindow("Blue Image", cv::WINDOW_AUTOSIZE);
        cv::namedWindow("Thres Image", cv::WINDOW_AUTOSIZE);
        cv::namedWindow("Erode Image", cv::WINDOW_AUTOSIZE);
        cv::namedWindow("Dilate Image", cv::WINDOW_AUTOSIZE);
        cv::namedWindow("Canny Image", cv::WINDOW_AUTOSIZE);
        cv::namedWindow("Lines Image", cv::WINDOW_AUTOSIZE);
      }

    private:
      // ROS parameters
      int blue_thres_;
      int erode_size_;
      int dilate_size_;
      int hough_rho_res_;
      double hough_theta_res_;
      int hough_threshold_;
      int hough_min_length_;
      int hough_max_gap_;

      rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr sub_image_;
      void recv_image(const sensor_msgs::msg::Image::ConstSharedPtr msg) {
        // Convert raw image from ROS image message into a cv::Mat
        cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, msg->encoding);
        cv::Mat raw_img = cv_ptr->image;

        cv::imshow("Raw Image", raw_img);
        cv::waitKey(1);

        // Split RGB image into its three separate channels
        std::vector<cv::Mat> split_images;
        cv::split(raw_img, split_images);

        // Extract the blue channel into its own grayscale image
        cv::Mat blue_image = split_images[0];

        cv::imshow("Blue Image", blue_image);
        cv::waitKey(1);

        // Apply binary threshold to create a binary image where white pixels correspond to high blue values
        cv::Mat thres_img;
        cv::threshold(blue_image, thres_img, blue_thres_, 255, cv::THRESH_BINARY);

        cv::imshow("Thres Image", thres_img);
        cv::waitKey(1);

        // Apply erosion to clean up noise
        cv::Mat erode_img;
        cv::erode(thres_img, erode_img, cv::Mat::ones(erode_size_, erode_size_, CV_8U));

        cv::imshow("Erode Image", erode_img);
        cv::waitKey(1);

        // Apply dilation to expand regions that passed the erosion filter
        cv::Mat dilate_img;
        cv::dilate(erode_img, dilate_img, cv::Mat::ones(dilate_size_, dilate_size_, CV_8U));

        cv::imshow("Dilate Image", dilate_img);
        cv::waitKey(1);

        // Apply Canny edge detection to reduce the number of points that are passed to Hough Transform
        cv::Mat canny_img;
        cv::Canny(dilate_img, canny_img, 1, 2);
        cv::imshow("Canny Image", canny_img);

        // Run Probabilistic Hough Transform algorithm to detect line segments
        std::vector<cv::Vec4i> line_segments;
        cv::HoughLinesP(canny_img, line_segments, hough_rho_res_, hough_theta_res_,
                        hough_threshold_, hough_min_length_, hough_max_gap_);

        // Draw detected Hough lines onto the raw image for visualization
        for (int i=0; i<line_segments.size(); i++){
          cv::line(raw_img, cv::Point(line_segments[i][0], line_segments[i][1]),
            cv::Point(line_segments[i][2], line_segments[i][3]), cv::Scalar(0, 0, 255));
        }

        cv::imshow("Lines Image", raw_img);
        cv::waitKey(1);
      }

      rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr param_cb_;
      rcl_interfaces::msg::SetParametersResult param_update(const std::vector<rclcpp::Parameter>& parameters) {
        for (const rclcpp::Parameter& p : parameters) {
          if (p.get_name() == "blue_thres") {
            blue_thres_ = p.as_int();
          } else if (p.get_name() == "erode_size") {
            erode_size_ = p.as_int();
          } else if (p.get_name() == "dilate_size") {
            dilate_size_ = p.as_int();
          } else if (p.get_name() == "hough_rho_res") {
            hough_rho_res_ = p.as_int();
          } else if (p.get_name() == "hough_theta_res") {
            hough_theta_res_ = p.as_double();
          } else if (p.get_name() == "hough_threshold") {
            hough_threshold_ = p.as_int();
          } else if (p.get_name() == "hough_min_length") {
            hough_min_length_ = p.as_int();
          } else if (p.get_name() == "hough_max_gap") {
            hough_max_gap_ = p.as_int();
          }
        }
        rcl_interfaces::msg::SetParametersResult result;
        result.successful = true;
        return result;
      }
  };
}

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  auto node_instance = std::make_shared<opencv_example::HoughTransformNode>();
  rclcpp::spin(node_instance);
  rclcpp::shutdown();
  return 0;
}